third_party/amd/lib/TritonAMDGPUTransforms/AccelerateAMDMatmul.cpp

```diff
@@ -406,6 +407,120 @@ Value convertAndCastTensor(PatternRewriter &rewriter, Value value,
   return castedTensor;
 }
 
+Value goThruScaleDecomposition(Operation *op, int opIdx) {
+  if (auto selectOp = dyn_cast<arith::SelectOp>(op)) {
+    op = selectOp.getFalseValue().getDefiningOp();
+  }
+
+  Value scale;
+  if (auto upcastFp8Op = dyn_cast<amdgpu::ScaledUpcastFp8Op>(op)) {
+    scale = upcastFp8Op.getScale();
+  } else if (auto upcastFp4Op = dyn_cast<amdgpu::ScaledUpcastFp4Op>(op)) {
+    scale = upcastFp4Op.getScale();
+  } else {
+    return {};
+  }
+
+  auto scaleTy = dyn_cast<RankedTensorType>(scale.getType());
+  if (!scaleTy)
+    return {};
+
+  // First check whether the scale is encoded as a BF16 value.
+  if (!scaleTy.getElementType().isBF16())
+    return {};
+
+  // Go through various shape manipulation ops.
+  tt::BitcastOp bitcastOp = nullptr;
+  while (scale) {
+    llvm::TypeSwitch<Operation *>(scale.getDefiningOp())
+        .Case<tt::BroadcastOp, tt::ExpandDimsOp, tt::ReshapeOp, tt::TransOp,
+              ttg::ConvertLayoutOp>([&](auto op) { scale = op.getSrc(); })
+        .Case<tt::BitcastOp>([&](auto op) { bitcastOp = op, scale = nullptr; })
+        .Default([&](auto op) { scale = nullptr; });
+  }
+  if (!bitcastOp)
+    return {};
+  auto shlOp = bitcastOp.getSrc().getDefiningOp<arith::ShLIOp>();
+  if (!shlOp)
+    return {};
+  auto extOp = shlOp.getLhs().getDefiningOp<arith::ExtUIOp>();
+  if (!extOp)
+    return {};
+  if (opIdx == 0)
+    return extOp.getIn();
+
+  auto transOp = extOp.getIn().getDefiningOp<tt::TransOp>();
+  if (!transOp)
+    return {};
+  return transOp.getSrc();
+}
+
+template <typename Op> Op getDefOpBeforeConvertLayout(Value op) {
+  while (auto cvtOp = op.getDefiningOp<ttg::ConvertLayoutOp>()) {
+    op = cvtOp.getSrc();
+  }
+  return op.getDefiningOp<Op>();
+}
+
+bool hasShuffledDotScale(Value v, int opIdx) {
```

Comment:
you shouldn't need to pattern match the shuffling. The idea of the shuffling is to have the user pick a HBM layout that will copy efficiently to what hardware supports. 
The compiler shouldn't have to detect it explicitly when we create the wmma it would set the native layout on the scale inputs. Then the layout propagation should propagate it all the way up to the load which should result in an efficient layout for the load operation. 
Therefore the compiler have no special path for it.

Reply:
Thanks for the comment! If my understanding is correct, right now we still need to identify the shuffling and manually set the layout. But according to https://github.com/triton-lang/triton/pull/7603#discussion_r2241387139, we have a ticket to track this and will remove this ad-hoc path eventually.

Reply:
> If my understanding is correct, right now we still need to identify the shuffling and manually set the layout.

do you know why? Any reason we can't get it right from the beginning?

Reply:
I put up https://github.com/triton-lang/triton/pull/8247 to make it not a hardcoded case pattern match.